retrieval/targetmanager.go

```diff
@@ -417,79 +417,96 @@ func providersFromConfig(cfg *config.ScrapeConfig) map[string]TargetProvider {
 	return providers
 }
 
-// targetsFromGroup builds targets based on the given TargetGroup and config.
-// Panics if target group is nil.
-func targetsFromGroup(tg *config.TargetGroup, cfg *config.ScrapeConfig) ([]*Target, error) {
-	targets := make([]*Target, 0, len(tg.Targets))
-
-	for i, labels := range tg.Targets {
-		for k, v := range cfg.Params {
-			if len(v) > 0 {
-				labels[model.LabelName(model.ParamLabelPrefix+k)] = model.LabelValue(v[0])
-			}
+// populateLabels builds a target from the given label set and scrape configuration.
+// Returns a nil target if the target is dropped due during relabeling.
+func populateLabels(lset model.LabelSet, cfg *config.ScrapeConfig) (res, orig model.LabelSet, err error) {
+	if _, ok := lset[model.AddressLabel]; !ok {
+		return nil, nil, fmt.Errorf("no address")
+	}
+	// Copy labels into the labelset for the target if they are not
+	// set already. Apply the labelsets in order of decreasing precedence.
+	scrapeLabels := model.LabelSet{
+		model.SchemeLabel:      model.LabelValue(cfg.Scheme),
+		model.MetricsPathLabel: model.LabelValue(cfg.MetricsPath),
+		model.JobLabel:         model.LabelValue(cfg.JobName),
+	}
+	for ln, lv := range scrapeLabels {
+		if _, ok := lset[ln]; !ok {
+			lset[ln] = lv
 		}
-		// Copy labels into the labelset for the target if they are not
-		// set already. Apply the labelsets in order of decreasing precedence.
-		labelsets := []model.LabelSet{
-			tg.Labels,
-			{
-				model.SchemeLabel:      model.LabelValue(cfg.Scheme),
-				model.MetricsPathLabel: model.LabelValue(cfg.MetricsPath),
-				model.JobLabel:         model.LabelValue(cfg.JobName),
-			},
+	}
+	// Encode scrape query parameters as labels.
+	for k, v := range cfg.Params {
+		if len(v) > 0 {
+			lset[model.LabelName(model.ParamLabelPrefix+k)] = model.LabelValue(v[0])
 		}
-		for _, lset := range labelsets {
-			for ln, lv := range lset {
-				if _, ok := labels[ln]; !ok {
-					labels[ln] = lv
-				}
-			}
+	}
+
+	preRelabelLabels := lset
+	lset = relabel.Process(lset, cfg.RelabelConfigs...)
+
+	// Check if the target was dropped.
+	if lset == nil {
+		return nil, nil, nil
+	}
+
+	// hasPort checks whether s is an IPv6 or IPv4 address that contains a port.
+	// We don't have to catch whether they are valid addresses, just whether they
+	// have a port if they are valid.
+	hasPort := func(s string) bool {
```

Comment:
Can `s` only be an IP address? What if it's a hostname (not sure if it can be at the point it's called)?